### PR TITLE
new branch to emphasize changes after adding the engine

### PR DESCRIPTION
**Pull request by b-artem**

### CODE REVIEW

**Review by Dulenchik — COMMENTED** [review]
@b-artem There are few things which look odd for me:

- dependencies in Gemfile duplicate gems in gemspec for engine;
- engine isn't independent: logic of application dances between engine and main app. IMO you should encapsulate it in engine. I believe that you also noticed that it's very uncomfortable to always reference to `main_app` or `shopping_cart`;

File: lib/engines/shopping_cart/lib/concerns/controllers/application_controller.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+module ShoppingCart
+  module Concerns
+    module Controllers
+      module ApplicationController
```
Comment (Dulenchik) [inline lib/engines/shopping_cart/lib/concerns/controllers/application_controller.rb]: @b-artem IMO this module should have another name, it's strange when you include it in `ApplicationController` in main app then you inherit engine's `ApplicationController` from main app `ApplicationController` and also include this module. 

Do you really need this module in both controllers or it's mistake?

File: lib/engines/shopping_cart/app/controllers/shopping_cart/application_controller.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+module ShoppingCart
+  class ApplicationController < ::ApplicationController
+    protect_from_forgery with: :exception
```
Comment (Dulenchik) [inline lib/engines/shopping_cart/app/controllers/shopping_cart/application_controller.rb]: @b-artem your engine should not override protection strategy

**Review by Dulenchik — APPROVED** [review]
@b-artem As I said before, engine depends from the main app too much, for example: method `authenticate_main_app_user!` exists in engine and refers to `authenticate_user!` it's very strong dependency. Or another example: method `main_app_products_url` expects that main app has `books_url` and I can't override this behavior (maybe I want to redirect on `root_path`)

When I said about duplications in gemfile - I mean `aasm`, as I checked - it's used only in engine's model.

Comments above are recommendations for the future. I'm ok to accept it, current implementation is good enough.

File: lib/engines/shopping_cart/app/controllers/shopping_cart/application_controller.rb (hunk 2 above)
Comment (b-artem) [reply lib/engines/shopping_cart/app/controllers/shopping_cart/application_controller.rb]: @Dulenchik Thank you, corrected it

File: lib/engines/shopping_cart/lib/concerns/controllers/application_controller.rb (hunk 1 above)
Comment (b-artem) [reply lib/engines/shopping_cart/lib/concerns/controllers/application_controller.rb]: @Dulenchik Thank you, that was a mistake. Corrected it. This concern is included in main app `ApplicationController` only now. Renamed it to `CartSupport` also.

**Comment by b-artem** [comment]
@Dulenchik Thank you, I'll try to make engines more independent in the future. As for `aasm`, I included it in gemfile intentionally, main app `review` model uses it.

As the engine heavily uses devise `authenticate_user!` method, how should I handle authentication in an engine next time to make it more independent?

**Comment by Dulenchik** [comment]
@b-artem look at the sidekiq gem for example, it doesn't provide any mechanisms to protect this info. You should think about your engine like about library and ask yourself:

- what your library should do? Is it solves specific problem or your gem looks like a swiss-knife?
- if you want to add authorization/authentication :
  - should you add a dependency from the some gems and implement hidden/encapsulated version inside engine;
  - expect for methods which main app should define for you;
  - inject a dependency via config in initializer;

which strategy you should choose? it depends on the requirements. If you build an open-source re-usable gem, probably you need the most flexible variant. If you extract a part to the engine, but it still lives in the same app/repository - the highest flexibility could be an overkill

**Pull request merged**